### PR TITLE
Feature/ball handling

### LINKED ISSUE #124: remove duplicate robothasball functions
`bool robothasball()` is at a lot of places.

- coach
- World
- controlUtils

we should merge these into one. (preferrably into the world one)

### LINKED ISSUE #126: Centralize ball visibility handling to world
Currently all over the place, needs to be in one place.

### PR DESCRIPTION
**Pull request by rolfvdhulst**
- [x] The tests are passing

### Summary
Does 2 things (which are heavily interwined):
- All ball visibility related handling to the world state: If a ball disappears whilst being dribbled the new location is assumed. Previously this was in several skills.
- Removes all the duplicate hasBall() related functions and created a single central one in the world class.

The currently used version has been tested with real robots and works decently well.

There are also some minor bugs which I fixed 
### Related issues
If there are related issues, reference them here.

fixes #124
fixes #126 

### Comments for the reviewer
Please look critically at all the skills. Since I had to change this in many places, I might have accidentally broken some functionality somewhere.

### Checklist for the reviewer
This acts as a guide to help you and/or the reviewer to analyze the code. 

**functionality**  
- Is the code is properly tested?
- Is the code working as expected
- Is the code is understandable and easy to read
- Are proper abstractions made? (no duplicate code, nice use of functions/objects)

**style**  
- Variables/objects/classes have proper names
- Proper use of variables: nu unused vars, no magic numbers
- No possible memory leaks. Mutexes properly used when multithreading? possible pointer issues?
- Are smart pointers used?
- unused variables?
- magic numbers?
- well documented?
- No unused imports?
- No leftover code / unused comments
- Proper indentation
- Proper use of namespaces

### CODE REVIEW

File: roboteam_ai/src/conditions/HasBall.cpp [hunk 1]
```diff
@@ -25,7 +25,7 @@ bt::Node::Status HasBall::update() {
     if (! robot || ! ball) {
         return Status::Failure;
     }
-    if (World::robotHasBall(*robot, *ball)) {
+    if (World::ourBotHasBall(robot->id)) {
```
Comment (RobZelluf) [inline roboteam_ai/src/conditions/HasBall.cpp]: I would suggest robotHasBall(robot->id, bool ourTeam)

File: roboteam_ai/src/skills/Attack.cpp [hunk 2]
```diff
@@ -36,7 +36,7 @@ bt::Node::Status Attack::onUpdate() {
         targetPos = ball;
         command.use_angle = 1;
         command.w = static_cast<float>(((Vector2) {- 1.0, - 1.0}*deltaBall).angle());
-        if (Coach::doesRobotHaveBall(robot->id, true, rtt::ai::Constants::MAX_BALL_RANGE())) {
+        if (World::ourBotHasBall(robot->id)) {
```
Comment (RobZelluf) [inline roboteam_ai/src/skills/Attack.cpp]: Change to general robotHasBall function

File: roboteam_ai/src/skills/Dribble.cpp [hunk 3]
```diff
@@ -11,7 +11,7 @@ Dribble::Dribble(string name, bt::Blackboard::Ptr blackboard)
 
 Dribble::Progression Dribble::checkProgression() {
     if (currentProgress == ON_THE_WAY) {
-        if (! robotHasBall()) {
+        if (! World::ourBotHasBall(robot->id,Constants::MAX_BALL_BOUNCE_RANGE())) {
```
Comment (RobZelluf) [inline roboteam_ai/src/skills/Dribble.cpp]: Is it really necessary to pass this constant on. Should the function be able to get it from Constants itself?

File: roboteam_ai/src/conditions/HasBall.cpp (hunk 1 above)
Comment (mrlukasbos) [reply roboteam_ai/src/conditions/HasBall.cpp]: agreed.

File: roboteam_ai/src/skills/Dribble.cpp (hunk 3 above)
Comment (rolfvdhulst) [reply roboteam_ai/src/skills/Dribble.cpp]: Well, yes, because we have 2 definitions of whether a robot has a ball or not.  (one at 0.05, with bounces allowed up until 0.15 m) This way we can also use this easily in the future for for example checking if a robot can kick a ball easily by passing a constants::MAX_KICKING_DIST or so.

File: roboteam_ai/src/conditions/HasBall.cpp (hunk 1 above)
Comment (rolfvdhulst) [reply roboteam_ai/src/conditions/HasBall.cpp]: Done

Comment (mrlukasbos) [reply roboteam_ai/src/conditions/HasBall.cpp]: No you still have ourBotHasBall and theirBotHasBall now

File: roboteam_ai/src/control/ControlUtils.cpp [hunk 4]
```diff
@@ -47,7 +47,7 @@ bool ControlUtils::pointInRectangle(Vector2 PointToCheck, Vector2 SP1, Vector2 S
     if (pointInTriangle(PointToCheck, SP1, SP2, SP3)) {
         return true;
     }
-    else return pointInTriangle(PointToCheck, SP4, SP1, SP2);
+    else return pointInTriangle(PointToCheck, SP4, SP1, SP3);
```
Comment (mrlukasbos) [inline roboteam_ai/src/control/ControlUtils.cpp]: why is this changed? was this a mistake?

File: roboteam_ai/src/utilities/World.cpp [hunk 5]
```diff
@@ -131,28 +134,95 @@ std::shared_ptr<roboteam_msgs::WorldBall> World::getBall() {
...
-}
-
-/// returns true if a robot at a given position and orientation has the ball
-bool World::robotHasBall(Vector2 robotPos, double robotOrientation, Vector2 ballPos, double frontDist) {
-    Vector2 dribbleLeft =
-            robotPos + Vector2(Constants::ROBOT_RADIUS(), 0).rotate(robotOrientation - Constants::DRIBBLER_ANGLE_OFFSET());
-    Vector2 dribbleRight =
-            robotPos + Vector2(Constants::ROBOT_RADIUS(), 0).rotate(robotOrientation + Constants::DRIBBLER_ANGLE_OFFSET());
 
+double World::findBallDist(roboteam_msgs::WorldRobot &bot, roboteam_msgs::WorldBall &ball) {
```
Comment (mrlukasbos) [inline roboteam_ai/src/utilities/World.cpp]: I would use the word 'get' instead of find'

File: roboteam_ai/src/utilities/World.cpp [hunk 6]
```diff
@@ -131,28 +134,95 @@ std::shared_ptr<roboteam_msgs::WorldBall> World::getBall() {
...
+// uses MAX_BALL_RANGE as default max Dist
+bool World::BotHasBall(int id, bool ourTeam, double maxDistToBall) {
+    if (ourTeam){
+        return ourBotHasBall(id,maxDistToBall);
+    }
+    else{
+        return theirBotHasBall(id,maxDistToBall);
+    }
+}
+bool World::ourBotHasBall(int id, double maxDistToBall){
```
Comment (mrlukasbos) [inline roboteam_ai/src/utilities/World.cpp]: you can shorten these functions. You could even make it one function: robotHasBall(int id, bool ourteam, double maxdisttoball)

File: roboteam_ai/src/utilities/World.cpp [hunk 7]
```diff
@@ -131,28 +134,95 @@ std::shared_ptr<roboteam_msgs::WorldBall> World::getBall() {
...
+bool World::theirBotHasBall(int id, double maxDistToBall){
+    std::lock_guard<std::mutex> lock(worldMutex);
+    if (TheirBotsBall.find(id)!=TheirBotsBall.end()){
+        if (TheirBotsBall[id]<=maxDistToBall){
+            return true;
+        }
+    }
+    return false;
+}
+// picks the bot that has the ball and is closest to it.
```
Comment (mrlukasbos) [inline roboteam_ai/src/utilities/World.cpp]: Why is this? is this for challenging robots or something??

File: roboteam_ai/src/utilities/World.cpp [hunk 8]
```diff
@@ -183,7 +253,58 @@ roboteam_msgs::World World::futureWorld(double time, double maxTimeOffset) {
...
+        }
+        for (auto bot: TheirBotsBall ) {
+            if (bot.second < maxDist) {
+                maxDist = bot.second;
+                bestId = bot.first;
+                ourTeam = false;
+            }
+        }
+        if (bestId==-1){
+            ROS_ERROR("WTF BALL HANDLING");
```
Comment (mrlukasbos) [inline roboteam_ai/src/utilities/World.cpp]: this is not professional.

File: roboteam_ai/src/utilities/World.cpp [hunk 9]
```diff
@@ -131,28 +134,95 @@ std::shared_ptr<roboteam_msgs::WorldBall> World::getBall() {
...
+int World::whichBotHasBall(bool ourTeam) {
+    std::lock_guard<std::mutex> lock(worldMutex);
+    double maxDist = 100;
+    int bestId = - 1;
+    if (ourTeam) {
+        for (auto bot: OurBotsBall) {
+            if (bot.second < maxDist) {
+                maxDist = bot.second;
+                bestId = bot.first;
+            }
```
Comment (mrlukasbos) [inline roboteam_ai/src/utilities/World.cpp]: this whole file feels like it has a lot of duplicates and it can be simplified quite a bit with different abstractions

File: roboteam_ai/src/control/ControlUtils.cpp (hunk 4 above)
Comment (rolfvdhulst) [reply roboteam_ai/src/control/ControlUtils.cpp]: See commit message, this was a very very ugly bug (it was not actually checking a rectangle).

File: roboteam_ai/src/conditions/HasBall.cpp (hunk 1 above)
Comment (rolfvdhulst) [reply roboteam_ai/src/conditions/HasBall.cpp]: should be fixed

File: roboteam_ai/src/jsons/strategies/testAvoidanceStategy.json [hunk 10]
```diff
@@ -0,0 +1,18 @@
+{
```
Comment (mrlukasbos) [inline roboteam_ai/src/jsons/strategies/testAvoidanceStategy.json]: why this strategy? what does it do? can it be removed?

File: roboteam_ai/src/conditions/HasBall.cpp (hunk 1 above)
Comment (RobZelluf) [reply roboteam_ai/src/conditions/HasBall.cpp]: Yes, this
